#! /bin/sh /usr/share/dpatch/dpatch-run
Author: Tim Retout <diocles@debian.org>, Christian Stimming <stimming@tuhh.de>
Origin: upstream, http://svn.gnucash.org/trac/changeset/19638
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=593479
Bug-Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=525549
Applied-Upstream: 2.4.0
Subject: Ensure not to accidentally delete our main account file

Original patch by Tim Retout who writes:

strptime is passed (name + pathlen + 1) as the string to search.  However, when
looking at the main account file, strlen(name) == pathlen, so strptime is
looking at the point just past the end of name.

Sometimes this will be parseable by strptime, and this leads to the account
file being unlinked.

@DPATCH@

diff --git a/src/backend/file/gnc-backend-file.c b/src/backend/file/gnc-backend-file.c
index da2386a..ab109a1 100644
--- a/src/backend/file/gnc-backend-file.c
+++ b/src/backend/file/gnc-backend-file.c
@@ -722,12 +722,20 @@ gnc_file_be_remove_old_files(FileBackend *be)
              continue;
 
         name = g_build_filename(be->dirname, dent, (gchar*)NULL);
-        len = strlen(name) - 4;
-
-        /* Is this file associated with the current data file */
-        if (strncmp(name, be->fullpath, pathlen) == 0) 
+        len = strlen(name);
+
+        /* Is this file associated with the current data file?
+	 * Additionally, the invariants for the pointer arithmetic
+         * must hold: String length long enough to contain the suffix,
+         * and string length large enough so that strptime below will
+         * not be passed a pointer outside of our string. (Otherwise
+         * the result of strptime might be parseable and the main data
+         * file is deleted, #593479) */
+        if ((strncmp(name, be->fullpath, pathlen) == 0)
+            && (len >= 4)
+            && (len > pathlen))
         {
-            if ((safe_strcmp(name + len, ".LNK") == 0) &&
+            if ((safe_strcmp(name + len - 4, ".LNK") == 0) &&
                 /* Is a lock file. Skip the active lock file */
                 (safe_strcmp(name, be->linkfile) != 0) &&
                 /* Only delete lock files older than the active one */
